include/seqan3/alignment/pairwise/alignment_algorithm.hpp

```diff
@@ -387,38 +363,66 @@ class alignment_algorithm :
                         second_range_t & second_range,
                         cache_t & cache)
     {
-        using std::get;
+        using zip_column_t = decltype(views::zip(*this->score_matrix_iter, *this->trace_matrix_iter));
+        using column_iter_t = std::ranges::iterator_t<decltype(std::declval<zip_column_t>() | views::take_exactly(1))>;
+
+        if (is_fst_range_empty)  // Already tracked the max in initialisation.
+            return;
+
         auto const & score_scheme = get<align_cfg::scoring>(*cfg_ptr).value;
-        std::ranges::for_each(first_range, [&, this](auto seq1_value)
-        {
-            // Move internal matrix to next column.
-            this->go_next_column();
 
-            auto col = this->current_column();
-            this->init_row_cell(*std::ranges::begin(col), cache);
+        column_iter_t column_iter{};
+        ranges::for_each(first_range, [&, this] (auto const seq1_value)
+        {
+            // Move to the next column in the alignment matrix and initialise first cell.
+            zip_column_t zip_column = views::zip(*++this->score_matrix_iter, *++this->trace_matrix_iter);
+            assert(!zip_column.empty());
 
-            auto second_range_it = std::ranges::begin(second_range);
-            std::ranges::for_each(col | ranges::view::drop_exactly(1), [&, this] (auto && cell)
+            // If the column is empty (the second range is empty)
+            if (is_snd_range_empty) // TODO [[unlikely]]
```

Comment:
You cannot say much about efficiency here. It would be probably very inefficient if the condition randomly changes and thus a jump is performed including flushing the instruction pipeline. But that condition is set outside of the loop and never changes inside. There is potentially a lot the compiler can do to optimise it. If we want to be sure it needs to be measured. But the performance optimisation is probably somethign I'll do in the background once we know the complete kernel including vectorisation.

Reply:
Ok, I sketch the implied changes here:
```cpp
if (is_snd_range_empty)
    std::for_each(rng, kernel_to_invoke(all parameters that are necessary));
else
    std::for_each(rng, other_kernel_to_invoke(all parameters that are necessary));
```

That means adding function call objects - maybe global, but preferably as nested structs of this class. That means repeating some logic inside of these or further refactoring that into other private functions. All of this needs to be guaranteed to be inlined, which might be very well the case. But all of it might change again once vectorisation is there and thoroughly benchmarked, since keeping all variables close to the executing kernel brings great performance benefits. And there were a lot of these unexpected things happening already in SeqAn. So I can or rather will change it if there are clear indications for performance boosts, otherwise I might going to change it if it clearly improves readability by keeping the performance stable. But all of it only happens when I am looking at the benchmarks.

Reply:
I know that what you are saying is completely right. In general, I completely agree with you. But I still would like to wait for final performance meassurements to come to the right conclusions and by adding hopefully the right means of abstractions to make the code still composable and efficient. Is that ok for the current state?